src/itensor/diagitensor.jl

```diff
@@ -0,0 +1,109 @@
+#
+# Diag ITensor constructors
+#
+
+"""
+    diagITensor([ElT::Type, ]v::Vector, inds...)
+    diagitensor([ElT::Type, ]v::Vector, inds...)
+
+Make a sparse ITensor with non-zero elements only along the diagonal.
+In general, the diagonal elements will be those stored in `v` and
+the ITensor will have element type `eltype(v)`, unless specified explicitly
+by `ElT`. The storage will have `NDTensors.Diag` type.
+
+In the case when `eltype(v) isa Union{Int, Complex{Int}}`, by default it will
+be converted to `float(v)`. Note that this behavior is subject to change
+in the future.
+
+The version `diagITensor` will never output an ITensor whose storage data
+is an alias of the input vector data.
+
+The version `diagitensor` might output an ITensor whose storage data
+is an alias of the input vector data in order to minimize operations.
+"""
+function diagITensor(
+  as::AliasStyle, elt::Type{<:Number}, v::AbstractVector{<:Number}, is::Indices
+)
+  length(v) ≠ mindim(is) && error(
+    "Length of vector for diagonal must equal minimum of the dimension of the input indices",
+  )
+  data = set_eltype(typeof(v), elt)(as, v)
+  return itensor(Diag(data), is)
+end
+
+function diagITensor(
+  as::AliasStyle, elt::Type{<:Number}, v::AbstractVector{<:Number}, is...
+)
+  return diagITensor(as, elt, v, indices(is...))
+end
+
+function diagITensor(as::AliasStyle, v::AbstractVector, is...)
+  return diagITensor(as, eltype(v), v, is...)
+end
+
+function diagITensor(as::AliasStyle, v::AbstractVector{<:RealOrComplex{Int}}, is...)
+  return diagITensor(as, float(eltype(v)), v, is...)
+end
+
+diagITensor(v::AbstractVector{<:Number}, is...) = diagITensor(NeverAlias(), v, is...)
+function diagITensor(elt::Type{<:Number}, v::AbstractVector{<:Number}, is...)
+  return diagITensor(NeverAlias(), elt, v, is...)
+end
+
+diagitensor(args...; kwargs...) = diagITensor(AllowAlias(), args...; kwargs...)
+
+# XXX TODO: explain conversion from Int
+# XXX TODO: proper conversion
+"""
+    diagITensor([elt::Type, ]x::Number, inds...)
+    diagitensor([elt::Type, ]x::Number, inds...)
+
+Make a sparse ITensor with non-zero elements only along the diagonal.
+In general, the diagonal elements will be set to the value `x` and
+the ITensor will have element type `eltype(x)`, unless specified explicitly
+by `ElT`. The storage will have `NDTensors.Diag` type.
+
+In the case when `x isa Union{Int, Complex{Int}}`, by default it will
+be converted to `float(x)`. Note that this behavior is subject to change
+in the future.
```

Comment:
I don't think this is the case if `elt` is specified explicitly, in that case `elt` will be taken literally. The automatic conversion should only happen if only `x` is specified, and is `Int` or `Complex{Int}`.

Reply:
@mtfishman I didn't change this comment or behavior of this code, I only moved the `diagITensor` to its own file see https://github.com/ITensor/ITensors.jl/blob/0a5a932eae91b5b8104ba6f19537003c6b17816f/src/itensor.jl#L447

Reply:
Ah ok! Even so, could you add a part to the end of that sentence like "...if an element type isn't specified explicitly."